### PR TITLE
Finalization of the Widget feature

### PR DESCRIPTION
**Pull request by TikhomirovSergey**
## Change list

- missed parameters of the `OverrideWidget` were added
  - iOSXCUITAutomation
  - windowsAutomation

- some improvements which are related to latest API changes were made.
- tests were refactored. Now it is supposed to use unit test for this feature
 
## Types of changes

- [ ] No changes in production code.
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

**Comment by TikhomirovSergey** [comment]
@SrinivasanTarget @mykola-mokhnach could you take a look at this PR?

**Comment by mykola-mokhnach** [comment]
It's huge O_O. Probably, this will take some time

**Comment by TikhomirovSergey** [comment]
@mykola-mokhnach. Ok. Don't hurry and be watchful.

File: src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/ExtendedWidgetTest.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+import static org.junit.Assert.assertThat;
+
+import org.junit.Test;
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebDriver;
+
+import java.util.List;
+
+public abstract class ExtendedWidgetTest extends WidgetTest {
+
```
Comment (mykola-mokhnach) [inline src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/ExtendedWidgetTest.java]: extra empty line

File: src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/StubWebElement.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    }
+
+    @Override
+    public WebDriver getWrappedDriver() {
+        return driver;
+    }
+
+    @Override
+    public String toString() {
+        return by.toString();
```
Comment (mykola-mokhnach) [inline src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/StubWebElement.java]: can _by_ be equal to null?

File: src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/android/AndroidWidgetTest.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+public class AndroidWidgetTest extends ExtendedWidgetTest {
+
+    public AndroidWidgetTest() {
+        super(new AndroidApp(), new AbstractStubWebDriver.StubAndroidDriver());
+    }
+
+    @Override
+    public void checkACommonWidget() {
+        defaultTest(app.getWidget(), app.getWidgets(),
```
Comment (mykola-mokhnach) [inline src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/android/AndroidWidgetTest.java]: can we have a better name for this method?

File: src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/combined/CombinedAppTest.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+public class CombinedAppTest extends WidgetTest {
+
+    /**
+     * Test data generation.
+     *
+     * @return test parameters
+     */
+    @Parameterized.Parameters
+    public static Collection<Object[]> data() {
+        return asList(new Object[][] {
```
Comment (mykola-mokhnach) [inline src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/combined/CombinedAppTest.java]: nit: Arrays.asList also supports simple objects (`T... item`) as parameters

File: src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/combined/CombinedWidgetTest.java [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+                equalTo(widgetClass));
+
+        List<Class<?>> classes = new ArrayList<>();
+
+        app.getWidgets().forEach(abstractWidget ->
+                classes.addAll(abstractWidget.getSubWidgets().stream()
+                        .map(widget -> widget.getSelfReference().getClass()).collect(toList())));
+
+        assertThat(classes,
+                contains(widgetClass, widgetClass, widgetClass, widgetClass));
```
Comment (mykola-mokhnach) [inline src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/combined/CombinedWidgetTest.java]: can you please explain what is being verified here?

File: src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/ios/IosApp.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+    @iOSFindBy(uiAutomator = "SOME_IOS_DEFAULT_LOCATOR")
+    @iOSXCUITFindBy(iOSNsPredicate = "SOME_XCUIT_DEFAULT_LOCATOR")
+    private DefaultIosWidget singleIosWidget;
+
+    @iOSFindBy(uiAutomator = "SOME_IOS_DEFAULT_LOCATOR")
+    @iOSXCUITFindBy(iOSNsPredicate = "SOME_XCUIT_DEFAULT_LOCATOR")
+    private List<DefaultIosWidget> multipleIosWidgets;
+
+    private AnnotatedIosWidget singleAnnotatedIosWidget;
```
Comment (mykola-mokhnach) [inline src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/ios/IosApp.java]: should these properties be annotated as well?

File: src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/windows/DefaultWindowsWidget.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+import io.appium.java_client.pagefactory_tests.widget.tests.AbstractWidget;
+import org.openqa.selenium.WebElement;
+
+import java.util.List;
+
+public class DefaultWindowsWidget extends AbstractWidget {
+
+    public static String WINDOWS_SUB_WIDGET_LOCATOR = "SOME_SUB_LOCATOR";
+
+    @WindowsFindBy(windowsAutomation =  "SOME_SUB_LOCATOR")
```
Comment (mykola-mokhnach) [inline src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/windows/DefaultWindowsWidget.java]: nit: format

File: src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/windows/WindowsApp.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+    @WindowsFindBy(windowsAutomation = "SOME_WINDOWS_DEFAULT_LOCATOR")
+    @iOSXCUITFindBy(iOSNsPredicate = "SOME_XCUIT_DEFAULT_LOCATOR")
+    private DefaultWindowsWidget singleIosWidget;
+
+    @WindowsFindBy(windowsAutomation = "SOME_WINDOWS_DEFAULT_LOCATOR")
+    @iOSXCUITFindBy(iOSNsPredicate = "SOME_XCUIT_DEFAULT_LOCATOR")
+    private List<DefaultWindowsWidget> multipleIosWidgets;
+
+    private AnnotatedWindowsWidget singleAnnotatedIosWidget;
```
Comment (mykola-mokhnach) [inline src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/windows/WindowsApp.java]: missing annotations?

File: src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/AbstractStubWebDriver.java [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+        public String getPlatformName() {
+            return ANDROID;
+        }
+
+        @Override
+        public String getAutomationName() {
+            return APPIUM;
+        }
+
+        public boolean isBrowser() {
```
Comment (mykola-mokhnach) [inline src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/AbstractStubWebDriver.java]: Override?

**Comment by TikhomirovSergey** [comment]
@mykola-mokhnach Ok. I'm working on you remarks.

File: src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/AbstractWidget.java [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+    public <T extends Widget> T getSubWidget() {
+        return null;
+    }
+
+    public <T extends Widget> List<T> getSubWidgets() {
+        return of();
+    }
+
+    public String toString() {
+        return getWrappedElement().toString();
```
Comment (TikhomirovSergey) [inline src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/AbstractWidget.java]: No

File: src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/android/AndroidWidgetTest.java [hunk 11]
```diff
@@ -17,26 +17,26 @@ public AndroidWidgetTest() {
     }
 
     @Override
-    public void checkACommonWidget() {
-        defaultTest(app.getWidget(), app.getWidgets(),
+    public void commonTestCase() {
+        testLogigByDefault(app.getWidget(), app.getWidgets(),
```
Comment (mykola-mokhnach) [inline src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/android/AndroidWidgetTest.java]: what is "Logig"?

File: src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/WidgetTest.java [hunk 12]
```diff
@@ -10,11 +11,15 @@
...
+    }
+
     protected WidgetTest(AbstractApp app, WebDriver driver) {
         this.app = app;
         initElements(new AppiumFieldDecorator(driver), app);
     }
 
     @Test
-    public abstract void checkACommonWidget();
+    public abstract void commonTestCase();
```
Comment (mykola-mokhnach) [inline src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/WidgetTest.java]: it is recommended having a verb in method names

File: src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/ios/IosWidgetTest.java [hunk 13]
```diff
@@ -17,26 +17,26 @@ public IosWidgetTest() {
...
-        defaultTest(app.getWidget(), app.getWidgets(),
+    public void commonTestCase() {
+        testLogigByDefault(app.getWidget(), app.getWidgets(),
                 IosUIAutomation(IOS_DEFAULT_WIDGET_LOCATOR), IosUIAutomation(IOS_SUB_WIDGET_LOCATOR));
     }
 
     @Override
-    public void checkAnAnnotatedWidget() {
-        defaultTest(((ExtendedApp) app).getAnnotatedWidget(), ((ExtendedApp) app).getAnnotatedWidgets(),
+    public void checkCaseWhenWidgetClassHasDeclaredLocatorAnnotation() {
```
Comment (mykola-mokhnach) [inline src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/ios/IosWidgetTest.java]: 👍

File: src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/android/AndroidWidgetTest.java (hunk 11 above)
Comment (TikhomirovSergey) [inline src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/android/AndroidWidgetTest.java]: It is typo :)

File: src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/StubWebElement.java (hunk 2 above)
Comment (TikhomirovSergey) [inline src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/StubWebElement.java]: @mykola-mokhnach I have fixed the code. Please take a look at the 
https://github.com/appium/java-client/pull/669/files/0db6a3093c6bb64f07e4f85c907da22452c293b1#diff-8d9d46cd20efb5a5c8a11fef4fc8e5e4R22

Also this situaation is not possible in production code

File: src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/ExtendedWidgetTest.java (hunk 1 above)
Comment (TikhomirovSergey) [inline src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/ExtendedWidgetTest.java]: @mykola-mokhnach FIXED

File: src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/AbstractStubWebDriver.java [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+    @Override
+    public abstract String getAutomationName();
+
+    @Override
+    public boolean isBrowser() {
+        return false;
+    }
+
+    @Override
+    public void get(String url) {
```
Comment (TikhomirovSergey) [inline src/test/java/io/appium/java_client/pagefactory_tests/widget/tests/AbstractStubWebDriver.java]: @mykola-mokhnach FIXED

**Pull request merged**